README.md
# Introduction
`minio-incinerator` (aka `mininc`) borrows heavily from [potman](https://github.com/bsdpot/potman), `clusterfurnace` and `cephsmelter` to build a virtualbox minio cluster .

Do not run in production! This is a testing environment to show minio running on FreeBSD, with Consul, and the Beast-of-Argh monitoring solution alongside.

# Outline
This will bring up 4 minio servers, configured for erasure coding, with a web interface:
* minio1 (8CPU, 4GB, 4 attached disks, nginx reverse proxy host, consul jail, beast-of-argh jail)
* minio2 (4CPU, 4GB, 4 attached disks)
* minio3 (4CPU, 4GB, 4 attached disks)
* minio4 (4CPU, 4GB, 4 attached disks)

# Requirements
The host computer running `minio-incinerator` needs at least 16 CPU threads, 24GB memory, plus 200GB disk space, preferably high speed SSD.

# Overview

## Quickstart
To create your own incinerator, init the VMs:

    git clone https://github.com/bretton/minio-incinerator.git
    cd minio-incinerator

      (edit) config.ini and set ACCESSIP to a free IP on LAN, and set DISKSIZE

    export PATH=$(pwd)/bin:$PATH
    (optional: sudo chmod 775 /tmp)
    mininc init mycluster
    cd mycluster
    mininc packbox
    mininc startvms
    ...
    mininc status
    ...

## Stopping

    mininc stopvms

## Destroying

    mininc destroyvms

## Dependencies

`minio-incinerator` requires
- ansible
- bash
- git
- packer
- vagrant
- virtualbox

# Installation and Operation

## FreeBSD

(todo: add steps from https://docs.freebsd.org/en/books/handbook/virtualization/#virtualization-host-virtualbox )

### Hosts which underwent an upgrade from 13.0 to 13.1
Please see the [ERRATA](ERRATA.md) file for info on fixing a broken upgrade.

### Install for FreeBSD

Install packages, start services and configure permissions and networks
```
pkg install bash git packer py39-ansible py39-packaging vagrant virtualbox-ose
service vboxnet enable
service vboxnet start
    
(sudo) pw groupmod vboxusers -m <username>

mkdir -p /usr/local/etc/vbox
vi /usr/local/etc/vbox/networks.conf
```

(add, with asterisk; this is extremely broad)
```
* 0.0.0.0/0
```

Add to /etc/rc.conf because the host needs to be a router for private networks in Virtualbox to get internet access
```
gateway_enable="YES"
```

Restart networking (may pause ssh session for a bit)
```
sudo service netif restart && sudo service routing restart
```

edit .profile, add the following, adjusting for username
```
PATH=/home/<username>/minio-incinerator/bin:$PATH; export PATH
```

Download and configure minio-incinerator:
```
git clone https://github.com/bretton/minio-incinerator
cd minio-incinerator

  (edit) config.ini and set ACCESSIP to a free IP on LAN, and set DISKSIZE

export PATH=$(pwd)/bin:$PATH
(optional: sudo chmod 775 /tmp)
mininc init -v mycluster
cd mycluster
mininc packbox
mininc startvms

  vagrant ssh minio1   
  vagrant ssh minio2
  vagrant ssh minio3
  vagrant ssh minio4
```

## Ubuntu 20.04 with Virtualbox
Install necessary packages
```
sudo apt-get install curl wget -y
sudo apt-get install ruby-full -y   # Ubuntu 20.04 is Ruby 2.7

sudo add-apt-repository ppa:git-core/ppa -y
sudo apt-get update
sudo apt-get install git -y

sudo apt-get install ansible virtualbox -y

curl -fsSL https://apt.releases.hashicorp.com/gpg | sudo apt-key add -
sudo apt-add-repository "deb [arch=amd64] https://apt.releases.hashicorp.com $(lsb_release -cs) main"
sudo apt-get update && sudo apt-get install packer vagrant
```

Download and configure minio-incinerator:
```
git clone https://github.com/bretton/minio-incinerator.git
cd minio-incinerator

  (edit) config.ini and set ACCESSIP to a free IP on LAN, and set DISKSIZE

export PATH=$(pwd)/bin:$PATH
mininc init -v mycluster
cd mycluster
mininc packbox
mininc startvms

  vagrant ssh minio1   
  vagrant ssh minio2
  vagrant ssh minio3
  vagrant ssh minio4
```

# Usage

    Usage: mininc command [options]

    Commands:
        destroyvms  -- Destroy VMs
        help        -- Show usage
        init        -- Initialize new minio-incinerator
        packbox     -- Create vm box image
        startvms    -- Start (and provision) VMs
        status      -- Show status
        stopvms     -- Stop VMs

## Minio Dashboard

The Minio web interface is available via nginx reverse proxy at https://ACCESSIP for the IP address set in ACCESSIP variable of `config.ini`.

Accept the self-signed certificate TWICE. First is for nginx, second minio. Remember, this is not a production-ready environment. 

Login with user `demoadmin` and `NP4c2KzyESKCIEsDk2I2Dmb2HAFsGSxec30Uxiqz` to access the dashboard.

You can now create a bucket and explore settings.

## config.ini

### Access IP

A virtual interface is created with a free IP address from the LAN. You must provide this free IP address in `config.ini` in the `ACCESSIP` section.

### Disk Sizes

Virtual disks are saved in the `minio-incinerator` directory for the duration of running the program. 

Set the size of the virtual disks in `config.ini` in the `DISKSIZE` section. 

Input disk size in thousands of MB, where 10000 (MB) is the same as 10GB. 

To set a 20GB disk, use 20000.

Do not add the metric M, or MB or GB! It won't work.

Sixteen (16) virtual drives will be created of this size, so ensure you have sufficient disk space!

A basic check is done to see if the input figures can be catered to, on init.



CHANGELOG.md
0.7

* Update for FreeBSD-13.1
* Inclusion of Beast-of-Argh system monitoring image

---

0.6

* Live on github
* README updates

---

0.5

* Cluster with 4 nodes, 4 drives each
* Boots with erasure coding config
* Updated README

---

0.4

* Generated certificate & key needs minio:minio ownership explicitly set
* Fixed nginx proxy, certificates need multiple IPs set
* Current version is 4 node cluster with 2 disks each, optimal would be 4 disks each

---

0.3

* Fixed bunch of bugs in previous version

---

0.2

* Use rsync instead of scp to copy SSL files between servers, else host verification errors
* Added lots of checks for ssh due to timeouts
* first working version with 2 pools, 2 servers each, 2 disks per server

---

0.1

* First bash at minio-incinerator, drawing inspiration from potman, clusterfurnace, cephsmelter and making adjustments
* Brings up 4 virtual machines, SSH connection issues syncing certificates around


commands/common.sh
#!/usr/bin/env bash

set -e

VERSION_REGEX='^[0-9](.[0-9a-zA-Z]+)*$'
ORIGIN_REGEX='^([a-zA-Z0-9_]*)$'
INCINERATOR_NAME_REGEX='^[a-zA-Z]([0-9a-zA-Z]+)*$'
FREEBSD_VERSION_REGEX='^(12\.2|13\.0|13\.1)$'
# poor
NETWORK_REGEX='^([0-9]{1,3})\.([0-9]{1,3})\.([0-9]{1,3})$'
# shellcheck disable=SC2034
FLAVOUR_REGEX='^[a-zA-Z][a-zA-Z0-9-]{0,19}[a-zA-Z0-9]$'

# Minio host vars
MINIO1=minio1
MINIO2=minio2
MINIO3=minio3
MINIO4=minio4

function common_init_vars() {
  STEPCOUNT=0
  STEP=
  case "$VERBOSE" in
    [Yy][Ee][Ss]|1)
      VERBOSE=1
    ;;
    *)
      VERBOSE=0
    ;;
  esac

  case "$DEBUG" in
    [Yy][Ee][Ss]|1)
      DEBUG=1
    ;;
    *)
      DEBUG=0
    ;;
  esac
}

function step() {
  ((STEPCOUNT+=1))
  STEP="$*"
  if [ -n "$LOGFILE" ]; then
    echo "$STEP" >> "$LOGFILE"
  fi
  [ $VERBOSE -eq 0 ] || echo "$STEPCOUNT. $STEP"
}

# Hacky, needs to be replaced
# shellcheck disable=SC2206 disable=SC2086 disable=SC2116
function read_ini_file() {
  OLD_IFS=$IFS
  ini="$(<$1)"                # read the file
  ini="${ini//[/\\[}"          # escape [
  ini="${ini//]/\\]}"          # escape ]
  IFS=$'\n' && ini=( ${ini} ) # convert to line-array
  ini=( ${ini[*]//;*/} )      # remove comments with ;
  ini=( ${ini[*]//#*/} )      # remove comments with #
  ini=( ${ini[*]/\	=/=} )  # remove tabs before =
  ini=( ${ini[*]/=\	/=} )   # remove tabs be =
  ini=( ${ini[*]/\ =\ /=} )   # remove anything with a space around =
  ini=( ${ini[*]/#\\[/\}$'\n'cfg_section_} ) # set section prefix
  ini=( ${ini[*]/%\\]/ \(} )    # convert text2function (1)
  ini=( ${ini[*]/=/=\( } )    # convert item to array
  ini=( ${ini[*]/%/ \)} )     # close array parenthesis
  ini=( ${ini[*]/%\\ \)/ \\} ) # the multiline trick
  ini=( ${ini[*]/%\( \)/\(\) \{} ) # convert text2function (2)
  ini=( ${ini[*]/%\} \)/\}} ) # remove extra parenthesis
  ini[0]="" # remove first element
  ini[${#ini[*]} + 1]='}'    # add the last brace

  for i in ${!ini[*]}; do
    if [[ ${ini[$i]} =~ ^([^=]+)=(.*$) ]]; then
      ini[$i]="config_${BASH_REMATCH[1]}=${BASH_REMATCH[2]}"
    fi
  done
  eval "$(echo "${ini[*]}")" # eval the result
  IFS=$OLD_IFS
}

# shellcheck disable=SC2154
function read_potman_config() {
  read_ini_file "$1"
  cfg_section_incinerator

  if [[ ! "${config_name}" =~ $INCINERATOR_NAME_REGEX ]]; then
      >&2 echo "invalid name in $1"
      exit 1
  fi

  if [[ "${config_vm_manager}" != "vagrant" ]]; then
      >&2 echo "invalid vm_manager in $1"
      exit 1
  fi

  if [[ ! "${config_freebsd_version}" =~ $FREEBSD_VERSION_REGEX ]]; then
    >&2 echo "unsupported freebsd version in $1"
    exit 1
  fi

  if [[ ! "${config_network}" =~ $NETWORK_REGEX ]]; then
    >&2 echo "invalid network in $1"
    exit 1
  fi
}

# shellcheck disable=SC2154
function read_flavour_config() {
  read_ini_file "$1"
  cfg_section_manifest

  if [ "$config_runs_in_nomad" != "true" ] &&
      [ "$config_runs_in_nomad" != "false" ]; then
    >&2 echo "invalid runs_in_nomad in manifest"
    exit 1
  fi

  if [[ ! "${config_version}" =~ $VERSION_REGEX ]]; then
      >&2 echo "invalid version in manifest"
      exit 1
  fi

  if [[ ! "${config_origin}" =~ $ORIGIN_REGEX ]]; then
      >&2 echo "invalid origin in manifest"
      exit 1
  fi

  if [ -z "$config_keep" ]; then
      config_keep=false
  fi

  if [ "$config_keep" != "true" ] &&
      [ "$config_keep" != "false" ]; then
    >&2 echo "invalid keep in manifest"
    exit 1
  fi
}

function main_usage() {
  echo "
Usage: $0 command

Commands:
    destroyvms  -- Destroy VMs
    help        -- Show usage
    init        -- Initialize new minio-incinerator
    packbox     -- Create vm box image
    startvms    -- Start (and provision) VMs
    status      -- Show status
    stopvms     -- Stop VMs
"
}

function exec_minio_incinerator() {
  CMD=$1
  shift
  exec \
    env INCLUDE_DIR="$(dirname "${BASH_SOURCE[0]}")" \
    env LOGFILE="${LOGFILE}" \
    "${INCLUDE_DIR}/${CMD}.sh" "$@"
}

function main() {
  set -e

  if [ $# -lt 1 ]; then
    main_usage
    exit 1
  fi

  CMD="$1"
  shift

  if [ "${CMD}" = "help" ]; then
    CMD="$1"
    if [ -z "$CMD" ]; then
      main_usage
      exit 0
    fi
    ARGS=("-h")
  else
    ARGS=("$@")
  fi

  if [ "${CMD}" = "init" ]; then
    LOGFILE=""
  else
    if [ ! -f potman.ini ]; then
      >&2 echo "Not inside an incinerator (no potman.ini found). Try 'mininc init'."
      exit 1
    fi

    read_potman_config potman.ini

    if [ "${PWD##*/}" != "${config_name}" ]; then
      >&2 echo "Incinerator name doesn't match directory name"
      exit 1
    fi
    LOGFILE="${PWD}/_build/$CMD.log"
  fi

  case "${CMD}" in
    destroyvms|init|packbox|startvms|status|stopvms)
       exec_minio_incinerator "${CMD}" "${ARGS[@]}"
      ;;
    *)
      main_usage
      exit 1
      ;;
  esac
}

function init_minio1_ssh() {
  SSHCONF_MINIO1="_build/.ssh_conf.$MINIO1"
  vagrant ssh-config "$MINIO1" > "$SSHCONF_MINIO1"
}

function init_minio2_ssh() {
  SSHCONF_MINIO2="_build/.ssh_conf.$MINIO2"
  vagrant ssh-config "$MINIO2" > "$SSHCONF_MINIO2"
}

function init_minio3_ssh() {
  SSHCONF_MINIO3="_build/.ssh_conf.$MINIO3"
  vagrant ssh-config "$MINIO3" > "$SSHCONF_MINIO3"
}

function init_minio4_ssh() {
  SSHCONF_MINIO4="_build/.ssh_conf.$MINIO4"
  vagrant ssh-config "$MINIO4" > "$SSHCONF_MINIO4"
}

function run_ssh_minio1 {
  if [ $DEBUG -eq 1 ]; then
    ssh -F "$SSHCONF_MINIO1" "$MINIO1" -- "$@" | tee -a "$LOGFILE"
    return "${PIPESTATUS[0]}"
  else
    ssh -F "$SSHCONF_MINIO1" "$MINIO1" -- "$@" >> "$LOGFILE"
  fi
}

function run_ssh_minio2 {
  if [ $DEBUG -eq 1 ]; then
    ssh -F "$SSHCONF_MINIO2" "$MINIO2" -- "$@" | tee -a "$LOGFILE"
    return "${PIPESTATUS[0]}"
  else
    ssh -F "$SSHCONF_MINIO2" "$MINIO2" -- "$@" >> "$LOGFILE"
  fi
}

function run_ssh_minio3 {
  if [ $DEBUG -eq 1 ]; then
    ssh -F "$SSHCONF_MINIO3" "$MINIO3" -- "$@" | tee -a "$LOGFILE"
    return "${PIPESTATUS[0]}"
  else
    ssh -F "$SSHCONF_MINIO3" "$MINIO3" -- "$@" >> "$LOGFILE"
  fi
}

function run_ssh_minio4 {
  if [ $DEBUG -eq 1 ]; then
    ssh -F "$SSHCONF_MINIO4" "$MINIO4" -- "$@" | tee -a "$LOGFILE"
    return "${PIPESTATUS[0]}"
  else
    ssh -F "$SSHCONF_MINIO4" "$MINIO4" -- "$@" >> "$LOGFILE"
  fi
}


commands/export.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

DATE=$(date "+%Y-%m-%d")

usage()
{
  echo "This packages a CURRENTLY RUNNING VirtualBox environment into a re-usable VirtualBox machine or Vagrant box."
  echo ""
  echo "Usage: mininc export [-hv]"
}

OPTIND=1
while getopts "hv" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
# wasn't running without this commented out
   usage
   exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
# shellcheck disable=SC1091
source "${INCLUDE_DIR}/common.sh"
common_init_vars

step "Load potman config"
read_potman_config potman.ini
# shellcheck disable=SC2154
FREEBSD_VERSION="${config_freebsd_version}"
FBSD="${FREEBSD_VERSION}"
FBSD_TAG=${FREEBSD_VERSION//./_}

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Make sure vagrant plugins are installed"
(vagrant plugin list | grep "vagrant-disksize" >/dev/null)\
  || vagrant plugin install vagrant-disksize

step "Checking if machines to export"
# read the modified output of VBoxManage into an array, one item per line
readarray -t mymachines <<< "$(VBoxManage list runningvms |sed -e 's|" {|,|g' -e 's|"||g' -e 's|}||g')"

# check the length of the array
len="${#mymachines[@]}"

# doing it this rudimentory way because not sure how else to do this
case "${len}" in
  0)
    echo "There are no machines to export. Exiting."
    exit 1
    ;;
  1)
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_name <<< "$(echo ${mymachines[0]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_friendly_name <<< "$(echo ${serverone_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverone_uuid <<< "$(echo ${mymachines[0]} | awk -F, '{print $2}')"
    ;;
  2)
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_name <<< "$(echo ${mymachines[0]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_friendly_name <<< "$(echo ${serverone_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverone_uuid <<< "$(echo ${mymachines[0]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_name <<< "$(echo ${mymachines[1]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_friendly_name <<< "$(echo ${servertwo_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read servertwo_uuid <<< "$(echo ${mymachines[1]} | awk -F, '{print $2}')"
    ;;
  3)
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_name <<< "$(echo ${mymachines[0]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_friendly_name <<< "$(echo ${serverone_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverone_uuid <<< "$(echo ${mymachines[0]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_name <<< "$(echo ${mymachines[1]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_friendly_name <<< "$(echo ${servertwo_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read servertwo_uuid <<< "$(echo ${mymachines[1]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverthree_name <<< "$(echo ${mymachines[2]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverthree_friendly_name <<< "$(echo ${serverthree_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverthree_uuid <<< "$(echo ${mymachines[2]} | awk -F, '{print $2}')"
    ;;
  4)
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_name <<< "$(echo ${mymachines[0]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverone_friendly_name <<< "$(echo ${serverone_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverone_uuid <<< "$(echo ${mymachines[0]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_name <<< "$(echo ${mymachines[1]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read servertwo_friendly_name <<< "$(echo ${servertwo_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read servertwo_uuid <<< "$(echo ${mymachines[1]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverthree_name <<< "$(echo ${mymachines[2]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverthree_friendly_name <<< "$(echo ${serverthree_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverthree_uuid <<< "$(echo ${mymachines[2]} | awk -F, '{print $2}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverfour_name <<< "$(echo ${mymachines[3]} | awk -F, '{print $1}')"
    # shellcheck disable=SC2162 disable=SC2086
    read serverfour_friendly_name <<< "$(echo ${serverfour_name} | awk -F_ '{print $2}')"
    # shellcheck disable=SC2034 disable=SC2162 disable=SC2086
    read serverfour_uuid <<< "$(echo ${mymachines[3]} | awk -F, '{print $2}')"
    ;;
   *)
    echo "Error, mininc is not configured for export of ${len} virtual machines"
    exit 1
    ;;
esac

step "Check if box files exist"
# TODO: check if output dir, currently /tmp, has any *.box files and remove them
# script will give error if existing .box files, no error handling for this yet

step "Package machines"
if [ -n "${serverone_name}" ]; then
    vagrant package --base "${serverone_name}" --output /tmp/"${serverone_friendly_name}".box
fi
if [ -n "${servertwo_name}" ]; then
    vagrant package --base "${servertwo_name}" --output /tmp/"${servertwo_friendly_name}".box
fi
if [ -n "${serverthree_name}" ]; then
    vagrant package --base "${serverthree_name}" --output /tmp/"${serverthree_friendly_name}".box
fi
if [ -n "${serverfour_name}" ]; then
    vagrant package --base "${serverfour_name}" --output /tmp/"${serverfour_friendly_name}".box
fi

# if DEBUG is enabled, dump the variables
if [ "$DEBUG" -eq 1 ]; then
    printf "\n\n"
    echo "Dump of variables"
    echo "================="
    echo "FBSD: $FBSD"
    echo "FBSD_TAG: $FBSD_TAG"
    echo "Version: $VERSION with suffix: $VERSION_SUFFIX"
    printf "\n\n"
    echo "Date: $DATE"
    printf "\n\n"
fi

step "Success"

echo "Don't forget to start VMs again with 'mininc startvms'!"


commands/status.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

DATE=$(date "+%Y-%m-%d")

usage()
{
  echo "Usage: mininc status [-hv]"
}

OPTIND=1
while getopts "hv" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
    usage
    exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
# shellcheck disable=SC1091
source "${INCLUDE_DIR}/common.sh"
common_init_vars

step "Load potman config"
read_potman_config potman.ini
# shellcheck disable=SC2154
FREEBSD_VERSION="${config_freebsd_version}"
FBSD="${FREEBSD_VERSION}"
FBSD_TAG=${FREEBSD_VERSION//./_}

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Make sure vagrant plugins are installed"
(vagrant plugin list | grep "vagrant-disksize" >/dev/null)\
  || VAGRANT_DISABLE_STRICT_DEPENDENCY_ENFORCEMENT=1 vagrant plugin install vagrant-disksize

step "Show vagrant status"
echo "
===> Vagrant status <==="
vagrant status | grep -E "($MINIO1|$MINIO2|$MINIO3|$MINIO4)"

# if DEBUG is enabled, dump the variables
if [ "$DEBUG" -eq 1 ]; then
    printf "\n\n"
    echo "Dump of variables"
    echo "================="
    echo "FBSD: $FBSD"
    echo "FBSD_TAG: $FBSD_TAG"
    echo "Version: $VERSION with suffix: $VERSION_SUFFIX"
    printf "\n\n"
    echo "Date: $DATE"
    printf "\n\n"
fi

step "Success"


commands/stopvms.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

usage()
{
  echo "Usage: mininc stopvms [-hv] [-m machine]"
}

MACHINES=()

OPTIND=1
while getopts "hvm:" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  m)
    MACHINES=( "${OPTARG}" )
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
    usage
    exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
# shellcheck disable=SC1091
source "${INCLUDE_DIR}/common.sh"
common_init_vars

mkdir -p _build

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Make sure vagrant plugins are installed"
(vagrant plugin list | grep "vagrant-disksize" >/dev/null)\
  || VAGRANT_DISABLE_STRICT_DEPENDENCY_ENFORCEMENT=1 vagrant plugin install vagrant-disksize

step "Stop vagrant vms"
vagrant halt "${MACHINES[@]}"

step "Success"


commands/startvms.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

usage()
{
  echo "Usage: mininc startvms [-hv] [-m machine]"
}

MACHINES=()

OPTIND=1
while getopts "hvm:" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  m)
    MACHINES+=( "${OPTARG}" )
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
    usage
    exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
# shellcheck disable=SC1091
source "${INCLUDE_DIR}/common.sh"
common_init_vars

mkdir -p _build

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Make sure vagrant plugins are installed"
(vagrant plugin list | grep "vagrant-disksize" >/dev/null)\
  || VAGRANT_DISABLE_STRICT_DEPENDENCY_ENFORCEMENT=1 vagrant plugin install vagrant-disksize

step "Start vagrant vms"
vagrant up --provision "${MACHINES[@]}"

step "Success"


commands/destroyvms.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

usage()
{
  echo "Usage: mininc destroyvms [-hv] [-m machine]"
}

MACHINES=()

OPTIND=1
while getopts "hvm:" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  m)
    MACHINES=( "${OPTARG}" )
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
    usage
    exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
source "${INCLUDE_DIR}/common.sh"
common_init_vars

mkdir -p _build

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Make sure vagrant plugins are installed"
(vagrant plugin list | grep "vagrant-disksize" >/dev/null)\
  || VAGRANT_DISABLE_STRICT_DEPENDENCY_ENFORCEMENT=1 vagrant plugin install vagrant-disksize

step "Destroy vagrant vms"
vagrant destroy -f "${MACHINES[@]}"

step "Success"


commands/packbox.sh
#!/usr/bin/env bash

if [ -z "$BASH_VERSION" ]; then
  >&2 echo "This needs to run in bash"
  exit 1
fi

usage()
{
  echo "Usage: mininc packbox [-hv]"
}

OPTIND=1
while getopts "hv" _o ; do
  case "$_o" in
  h)
    usage
    exit 0
    ;;
  v)
    # shellcheck disable=SC2034
    VERBOSE="YES"
    ;;
  *)
    usage
    exit 1
    ;;
  esac
done

shift "$((OPTIND-1))"

if [ $# -ne 0 ]; then
  usage
  exit 1
fi

set -eE
trap 'echo error: $STEP failed' ERR
# shellcheck disable=SC1091
source "${INCLUDE_DIR}/common.sh"
common_init_vars

step "Start"
mkdir -p _build
read_potman_config potman.ini
# shellcheck disable=SC2154
FREEBSD_VERSION="${config_freebsd_version}"

step "Check tooling"
ansible --version >/dev/null
git --version >/dev/null
packer --version >/dev/null
vagrant --version >/dev/null
vboxheadless --version >/dev/null

step "Check box already exists"
if vagrant box list | grep "FreeBSD-${FREEBSD_VERSION}-RELEASE-amd64" |\
	  grep "virtualbox" >/dev/null; then
  step "Box already exists"
  exit 0
fi

step "Initialize"

rm -rf _build/packer
git clone -b testing https://github.com/bretton/packer-FreeBSD.git _build/packer
cd _build/packer

# future-proofing but simply replaces 13.1 with 13.1 currently
<variables.json.sample sed -e "s|13.1|${FREEBSD_VERSION}|g" >variables.json
#<variables.json.sample sed -e "s|13.0|${FREEBSD_VERSION}|g" -e "s|32G|10240|g" >variables.json
#cp -f variables.json.sample variables.json

#step "Validate packer build files"
#packer validate -only=virtualbox-iso template.json

step "Packer build"
# debug
export PACKER_LOG=1
export PACKER_BUILDER_TYPE="virtualbox-iso"
if [ -f variables.json ]; then
    packer build -only="virtualbox-iso" -var-file="variables.json" template.json
fi
# packer build -only=virtualbox-iso -var-file=variables.json template.json

step "Add Vagrant box"
vagrant box add "builds/FreeBSD-${FREEBSD_VERSION}-RELEASE-amd64.box" \
  --name "FreeBSD-${FREEBSD_VERSION}-RELEASE-amd64"

step "Success"
